litmus-portal/graphql-server/utils/misc.go

```diff
@@ -31,55 +31,55 @@ func RandomString(n int) string {
-	if cluster.AgentNsExists == nil {
-		cluster.AgentNsExists = &defaultState
-	}
-	if cluster.AgentSaExists == nil {
-		cluster.AgentSaExists = &defaultState
-	}
+	var (
+		lines                     []string
+		defaultState              = false
+		AgentNamespace            string
+		ServiceAccountName        string
+		DefaultAgentNamespace     = "litmus"
+		DefaultServiceAccountName = "litmus"
+	)
+
 	file, err := os.Open(template)
 	if err != nil {
 		return []byte{}, err
 	}
 	defer file.Close()
 	scanner := bufio.NewScanner(file)
-	var lines []string
 
-	if !*cluster.AgentNsExists || !*cluster.AgentSaExists {
-		comment := fmt.Sprintf("# This is an auto-generated file. DO NOT EDIT")
-		lines = append(lines, comment)
+	if cluster.AgentNsExists == nil {
+		cluster.AgentNsExists = &defaultState
 	}
-	if !*cluster.AgentNsExists {
-		var nsYaml string
-		if cluster.AgentNamespace != nil {
-			nsYaml = fmt.Sprintf("apiVersion: v1\nkind: Namespace\nmetadata:\n  name: %s\n---", *cluster.AgentNamespace)
-		} else {
-			nsYaml = fmt.Sprintf("apiVersion: v1\nkind: Namespace\nmetadata:\n  name: litmus\n---")
-		}
-		lines = append(lines, nsYaml)
+	if cluster.AgentSaExists == nil {
+		cluster.AgentSaExists = &defaultState
 	}
-	if !*cluster.AgentSaExists {
-		var saYaml string
-		if cluster.Serviceaccount != nil && cluster.AgentNamespace != nil {
-			saYaml = fmt.Sprintf("apiVersion: v1\nkind: ServiceAccount\nmetadata:\n  name: %s\n  namespace: %s\n---", *cluster.Serviceaccount, *cluster.AgentNamespace)
-		} else if cluster.Serviceaccount == nil && cluster.AgentNamespace != nil {
-			saYaml = fmt.Sprintf("apiVersion: v1\nkind: ServiceAccount\nmetadata:\n  name: litmus\n  namespace: %s\n---", *cluster.AgentNamespace)
-		} else if cluster.Serviceaccount != nil && cluster.AgentNamespace == nil {
-			saYaml = fmt.Sprintf("apiVersion: v1\nkind: ServiceAccount\nmetadata:\n  name: %s\n  namespace: litmus\n---", *cluster.Serviceaccount)
-		} else {
-			saYaml = fmt.Sprintf("apiVersion: v1\nkind: ServiceAccount\nmetadata:\n  name: litmus\n  namespace: litmus\n---")
-		}
+
+	if !*cluster.AgentNsExists && cluster.AgentNamespace != nil && *cluster.AgentNamespace != "" {
+		AgentNamespace = *cluster.AgentNamespace
+	} else {
+		AgentNamespace = DefaultAgentNamespace
+	}
+
+	if !*cluster.AgentSaExists && cluster.Serviceaccount != nil && *cluster.Serviceaccount != "" {
+		ServiceAccountName = *cluster.Serviceaccount
+	} else {
+		ServiceAccountName = DefaultServiceAccountName
+	}
+
+	var (
+		headerStr         = "# This is an auto-generated file. DO NOT EDIT\n"
+		namspaceStr       = "\napiVersion: v1\nkind: Namespace\nmetadata:\n  name: " + AgentNamespace + "\n"
+		serviceAccountStr = "---\napiVersion: v1\nkind: ServiceAccount\nmetadata:\n  name: " + ServiceAccountName + "\n  namespace: " + AgentNamespace + "\n---"
+	)
+
+	if cluster.AgentNsExists == nil && *cluster.AgentNsExists == false && cluster.AgentSaExists != nil && *cluster.AgentSaExists == false {
```

Comment:
Typo - `cluster.AgentNsExists != nil`

Reply:
Also we need to separate the ns and sa yaml addition based on each of their existence. For instance - what will happen if the namespace exists while the service account doesn't.

Reply:
Yes, Actually we don't need `cluster.AgentNsExists != nil && cluster.AgentSaExists != nil` because it was checking before coming to this statement.  If it is nil, the default value is false.

Reply:
https://github.com/litmuschaos/litmus/blob/e6ee276762d4e96761cd593c4737e96994a1d3fa/litmus-portal/graphql-server/utils/misc.go#L50